cp2k-wrapper.sh

```diff
@@ -0,0 +1,168 @@
+
+
+# set correct temporary directory
+if [[ -z "${TMPDIR}" ]]
+then
+    TMPDIR="/tmp"
+fi
+
+
+# get list of allocated nodes
+HOSTFILE="${TMPDIR}/hostfile.${SLURM_JOB_ID}"
+srun hostname -s | sort | uniq -c | awk '{print $2" slots="$1}' > "${HOSTFILE}" || { rm -f "${HOSTFILE}"; exit ${E_WR_HOSTFILE}; }
+
+
+# print short summary
+print_summary ${ID} ${RUNTIME} ${PARTITION} "${DATAROOT}" ${SLURM_JOB_NUM_NODES}
+echo
+echo
+
+
+# set correct number of cores per node
+declare -i NUMCORES
+case "${PARTITION,,}" in
+    test|compute)
+        NUMCORES=14
+        ;;
+
+    pascal)
+        NUMCORES=12
+        ;;
+
+    *)
+        NUMCORES=1
+        ;;
+esac
+
+
+# distribute nodes between tasks accordingly and run them
+declare -i node
+node=1
+
+declare -i tnum
+
+for ((tnum=1; tnum <= NUMTASKS; tnum++))
+do
+    # read task line from runlist
+    line=`sed -n "${tnum},${tnum}p" "${DATAROOT}/runlist.${ID}"`
+
+    # remove preceding spaces
+    line=$(chomp "${line}")
+
+    # get nodes, threads, data directory and prepare nodelist for command execution
+    DATADIR=$(chomp "`echo "${line}" | awk '{$1 = ""; $2 = ""; print $0}'`")
+    cd "${DATADIR}"
+
+    NUMNODES=`echo "${line}" | awk '{print $1}'`
+    NUMTHREADS=`echo "${line}" | awk '{print $2}'`
+    NODELIST=`sed -n "${node},$((node + NUMNODES - 1))p" "${HOSTFILE}"`
+    let "node += NUMNODES"
+
+    echo "${NODELIST}" > hostfile.${ID}
+
+    # get command to run
+    COMMAND=`cat "runcmd.${ID}"`
+
+    # short summary for current task
+    echo "Data directory is [${DATADIR}]"
+    echo "Allocated nodes are:"
+    echo "${NODELIST}" | awk '{print $1}'
+    echo "Command to run is [${COMMAND}]"
+    echo
+
+    # construct final run command depending on executable filename, working partition and threads number
+    RUNCMD=""
+
+    case $(binname "${COMMAND}") in
+        cp2k.ssmp)
+            NODELIST=`echo "${NODELIST}" | awk '{print $1}'` # leave only node hostname
+            RUNCMD="export MKL_NUM_THREADS=${NUMCORES}; export OMP_NUM_THREADS=${NUMCORES};srun --nodes=1 --nodelist=${NODELIST} ${COMMAND}"
+            ;;
+
+        cp2k.popt)
+            check_exec ${L2_PRINT_LOG} "mpirun"
+
+            sed -i "s/slots=1/slots=${NUMCORES}/g" hostfile.${ID}
+
+            if [[ "${NUMTHREADS}" -ne 0 ]]
+            then
+                RUNCMD="mpirun --hostfile hostfile.${ID} -np ${NUMTHREADS} --nooversubscribe ${COMMAND}"
+            else
+                RUNCMD="mpirun --hostfile hostfile.${ID} --npernode ${NUMCORES} --nooversubscribe ${COMMAND}"
+            fi
+            ;;
+
+        cp2k.psmp)
+            check_exec ${L2_PRINT_LOG} "mpirun"
+			#For PSMP version default OMP_NUM_THREADS=2, adjustment to specific job may be effective
+			OMP_NUM_THREADS=2
+			#let EFFECTIVECORES=NUMCORES
+			let EFFECTIVECORES=NUMCORES/OMP_NUM_THREADS
+			
+			sed -i "s/slots=1/slots=${EFFECTIVECORES}/g" hostfile.${ID}
+			
+            if [[ "${PARTITION,,}" == "pascal" ]]
```

Comment:
Can we move all specific options for this run (`--mca mpi_warn_on_fork 0 --mca btl_openib_eager_limit 65536 --mca btl_openib_max_eager_rdma 8  --mca btl_openib_eager_rdma_num 8`) move to temporary variable just before `if` block to prevent code duplication?

Reply:
Moved into new variable $MCA_PARAMS